pandas/core/reshape/melt.py

```diff
@@ -72,7 +72,14 @@ def melt(
                     "The following 'value_vars' are not present in "
                     f"the DataFrame: {list(missing)}"
                 )
-        frame = frame.loc[:, id_vars + value_vars]
+        # use `iloc` instead of `loc` when `col_level` is specified
+        if col_level is not None:
+            nlevels = frame.columns.nlevels
+            iid_vars = [int(cols.index(elm) / nlevels) for elm in id_vars]
+            ivalue_vars = [int(cols.index(elm) / nlevels) for elm in value_vars]
```

Comment:
Thanks @ashtou 

Would it work to do something like
```python
idx = frame.columns.get_level_values(col_level).get_indexer(id_vars+value_vars)
frame = frame.iloc[:, idx]
```
?

Reply:
@MarcoGorelli of course! thanks for the tip! passes the tests too.